File: java/src/main/java/toolgood/algorithm/internals/functions/datetimes/Function_WORKDAY.java
package toolgood.algorithm.internals.functions.datetimes;

import java.util.HashSet;
import java.util.List;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_N;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.operands.MyDate;

public final class Function_WORKDAY extends Function_N {
    public Function_WORKDAY(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public String Name() {
        return "Workday";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        Operand args1 = GetDate(engine, tempParameter, 0);
        if (args1.IsErrorOrNone()) {
            return args1;
        }

        Operand args2 = GetNumber(engine, tempParameter, 1);
        if (args2.IsErrorOrNone()) {
            return args2;
        }

        MyDate startMyDate = args1.DateValue();
        int days = args2.IntValue();
        HashSet<String> list = new HashSet<>();

        for (int i = 2; i < funcs.length; i++) {
            Operand ar = GetDate(engine, tempParameter, i);
            if (ar.IsErrorOrNone()) {
                return ar;
            }
            MyDate holiday = ar.DateValue();
            list.add(holiday.Year + "-" + holiday.Month + "-" + holiday.Day);
        }

        while (days > 0) {
            startMyDate = startMyDate.AddDays(1);
            int dayOfWeek = startMyDate.DayOfWeek();
            if (dayOfWeek == 1 || dayOfWeek == 7) {
                continue;
            }
            String dateStr = startMyDate.Year + "-" + startMyDate.Month + "-" + startMyDate.Day;
            if (list.contains(dateStr)) {
                continue;
            }
            days--;
        }
        return Operand.Create(startMyDate);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.DATE;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType,
            String op, String val) {
        funcs[0].GetParameterTypes(noneEngine, result, OperandType.DATE);
        funcs[1].GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        for (int i = 2; i < funcs.length; i++) {
            funcs[i].GetParameterTypes(noneEngine, result, OperandType.DATE);
        }
    }
}

File: JavaScript/src/Internals/Functions/DateTimes/Function_WORKDAY.js
import { Function_N } from '../Function_N.js';
import { MyDate } from '../../MyDate.js';
import { Operand } from '../../../Operand.js';


class Function_WORKDAY extends Function_N {
    get Name() {
        return "Workday";
    }

    constructor(z) {
        super(z);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getDate(engine, tempParameter, 0);
        if (args1.IsError) { return args1; }

        let args2 = this.getNumber(engine, tempParameter, 1);
        if (args2.IsError) { return args2; }

        let startMyDate = new Date(args1.DateValue.ToDateTime().getTime());
        let days = args2.IntValue;
        let list = new Set();
        for (let i = 2; i < this.z.length; i++) {
            let ar = this.getDate(engine, tempParameter, i);
            if (ar.IsError) { return ar; }
            // 将日期转换为YYYY-MM-DD格式以确保Set能够正确比较
            let dateStr = ar.DateValue.ToDateTime().toISOString().split('T')[0];
            list.add(dateStr);
        }

        // 判断是否周末
        let isWeekend = (d) => {
            let dow = d.getDay();
            return dow === 0 || dow === 6;
        };
        // 仅取日期部分（当天00:00）用于比较
        let dateOnly = (d) => new Date(d.getFullYear(), d.getMonth(), d.getDate());

        if (days > 0) {
            // 先逐天对齐到周一（最多 6 天），期间消耗工作日
            while (startMyDate.getDay() !== 1 && days > 0) {
                startMyDate.setDate(startMyDate.getDate() + 1);
                if (isWeekend(startMyDate)) continue;
                if (list.has(startMyDate.toISOString().split('T')[0])) continue;
                days--;
            }
            if (days === 0) return Operand.Create(new MyDate(startMyDate));

            // 整周粗跳：起点已是周一，每 5 个工作日 = 7 天
            let afterJump = new Date(startMyDate.getTime());
            afterJump.setDate(afterJump.getDate() + Math.floor(days / 5) * 7);
            let extra = 0;
            let s0 = dateOnly(startMyDate);
            let j0 = dateOnly(afterJump);
            for (let h of list) {
                let hDate = new Date(h + 'T00:00:00');
                if (dateOnly(hDate) > s0 && dateOnly(hDate) <= j0 && !isWeekend(hDate)) {
                    extra++;
                }
            }
            startMyDate = afterJump;
            days = (days % 5) + extra;
            while (days > 0) {
                startMyDate.setDate(startMyDate.getDate() + 1);
                if (isWeekend(startMyDate)) continue;
                if (list.has(startMyDate.toISOString().split('T')[0])) continue;
                days--;
            }
        } else if (days < 0) {
            // 先逐天对齐到周五（最多 6 天），期间消耗工作日
            while (startMyDate.getDay() !== 5 && days < 0) {
                startMyDate.setDate(startMyDate.getDate() - 1);
                if (isWeekend(startMyDate)) continue;
                if (list.has(startMyDate.toISOString().split('T')[0])) continue;
                days++;
            }
            if (days === 0) return Operand.Create(new MyDate(startMyDate));

            // 整周粗跳：起点已是周五，每 5 个工作日 = 7 天
            let afterJump = new Date(startMyDate.getTime());
            afterJump.setDate(afterJump.getDate() + Math.floor(-days / 5) * -7);
            let extra = 0;
            let s0 = dateOnly(startMyDate);
            let j0 = dateOnly(afterJump);
            for (let h of list) {
                let hDate = new Date(h + 'T00:00:00');
                if (dateOnly(hDate) >= j0 && dateOnly(hDate) < s0 && !isWeekend(hDate)) {
                    extra++;
                }
            }
            startMyDate = afterJump;
            days = -((-days) % 5) - extra;
            while (days < 0) {
                startMyDate.setDate(startMyDate.getDate() - 1);
                if (isWeekend(startMyDate)) continue;
                if (list.has(startMyDate.toISOString().split('T')[0])) continue;
                days++;
            }
        }
        return Operand.Create(new MyDate(startMyDate));
    }
}

export { Function_WORKDAY };

